### PR TITLE
Sa 3471 admu disable scheduled tasks

### PR DESCRIPTION
**Pull request by kmaranionjc**
## Issues
* [SA-3471](https://jumpcloud.atlassian.net/browse/<jira-id>) - SA-3471-disable-scheduled-stasks

## What does this solve?
Some scheduled tasks prevents migrations from occuring. This change disables tasks that are on "Ready" state before the migration to ensure migration runs without tasks preventing it.
## Is there anything particularly tricky?
Checking if the tasks are disabled in the middle of the migration
## How should this be tested?
1. Open two PWSH terminal, for migration and task check
2. Run this command to check the pre-migration tasks that are not Windows and Ready: `Get-ScheduledTask | Where-Object { $_.TaskPath -notlike "*\Microsoft\Windows*" -and $_.State -ne "Disabled" -and $_.state -ne "Running" }`
3. Run ADMu
4. In the middle of migration, check if the tasks from pre-migration check are disabled: `Get-ScheduledTask | Where-Object { $_.TaskPath -notlike "*\Microsoft\Windows*" -and $_.State -eq "Disabled" }`
5. Lastly, after the migration, check if pre-migration tasks are set to "Ready": `Get-ScheduledTask | Where-Object { $_.TaskPath -notlike "*\Microsoft\Windows*" -and $_.State -ne "Disabled" -and $_.state -ne "Running" }`

## Screenshots
Pre:
image

Mid and after migration:
image

[SA-3471]: https://jumpcloud.atlassian.net/browse/SA-3471?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: jumpcloud-ADMU/Powershell/Tests/PSScriptAnalyzerSettings.psd1 [hunk 1]
```diff
@@ -32,15 +32,16 @@
...
-                    )
+        'PSAvoidUsingWriteHost',
+        'PSAvoidUsingConvertToSecureStringWithPlainText'
+        'PSAvoidUsingPlainTextForPassword',
+        'PSAvoidUsingUsernameAndPasswordParams',
+        'PSAvoidUsingWMICmdlet',
+        'PSAvoidUsingInvokeExpression',
+        'PSUseDeclaredVarsMoreThanAssignments',
+        'PSUseBOMForUnicodeEncodedFile',
+        'PSAvoidGlobalVars'
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/Tests/PSScriptAnalyzerSettings.psd1:43]: I've added `PSAvoidGlobalVairables` so we can print out the test job names while we are running the tests. The logs were getting too difficult to parse out and debug.

File: jumpcloud-ADMU/Powershell/Tests/Migration.Tests.ps1 [hunk 2]
```diff
@@ -1,3 +1,18 @@
...
+                $GLOBAL:TestName = $Context.Test.Name
+            }
+            EachTestTeardown   = {
+                $GLOBAL:TestName = $null
+            }
+        }
+        $state.Plugin += New-PluginObject @PluginParams
+    }
+}
+
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/Tests/Migration.Tests.ps1:15]:
```suggestion
# Add test name as variable
# https://github.com/pester/Pester/issues/1611
```

File: jumpcloud-ADMU/Powershell/Tests/Migration.Tests.ps1 [hunk 3]
```diff
@@ -44,6 +64,51 @@ Describe 'Migration Test Scenarios' {
             }
         }
 
+        It "Tests for the Scheduled Task already enabled" {
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/Tests/Migration.Tests.ps1:67]:
```suggestion
        It "Tests that a previously enabled Scheduled Task is enabled at the end of user migration" {
```

File: jumpcloud-ADMU/Powershell/Tests/Migration.Tests.ps1 [hunk 4]
```diff
@@ -44,6 +64,51 @@ Describe 'Migration Test Scenarios' {
...
+            { Start-Migration -AutobindJCUser $false -JumpCloudUserName $user2 -SelectedUserName "$ENV:COMPUTERNAME\$user1" -TempPassword "$($Password)" } | Should -Not -Throw
+
+            # Get the scheduled task
+            $task = Get-ScheduledTask -TaskName "TestTask"
+            # Task state should be ready
+            $task.State | Should -Be "Ready"
+
+
+        }
+        It "Tests for the Scheduled Task disabled" {
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/Tests/Migration.Tests.ps1:90]:
```suggestion
        It "Tests that a previously disable Scheduled Task is disabled after migration" {
```

File: jumpcloud-ADMU/Powershell/Tests/Migration.Tests.ps1 [hunk 5]
```diff
@@ -94,51 +159,140 @@ Describe 'Migration Test Scenarios' {
...
-                        [string]
-                        $Password,
-                        [Parameter()]
-                        [string]
-                        $JCUserName
-                    )
-                    $file = "C:\Users\$JCUserName"
-                    # wait for the new user
-                    while (!(Test-Path -Path $file -ErrorAction SilentlyContinue)) {
+        It "Tests that the tool can recover when the start migration script fails" {
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/Tests/Migration.Tests.ps1:162]:
```suggestion
        It "Tests that the tool can recover when the start migration script fails and that scheduled tasks are returned to their previous state" {
```

File: jumpcloud-ADMU/Powershell/Start-Migration.ps1 [hunk 6]
```diff
@@ -1401,7 +1401,7 @@ Function Start-Migration {
     Begin {
         Write-ToLog -Message:('####################################' + (get-date -format "dd-MMM-yyyy HH:mm") + '####################################')
         # Start script
-        $admuVersion = '2.4.2'
+        $admuVersion = '2.4.4'
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/Start-Migration.ps1:1404]: Should our version for this module should be 3.5.0? I'd argue that this is new functionality, not solving a bug or addressing a previous issue.

File: jumpcloud-ADMU/Powershell/Start-Migration.ps1 [hunk 7]
```diff
@@ -1546,6 +1546,28 @@ Function Start-Migration {
             new-item -ItemType Directory -Force -Path $jcAdmuTempPath 2>&1 | Write-Verbose
         }
         Write-ToLog -Message:($localComputerName + ' is currently Domain joined to ' + $WmiComputerSystem.Domain + ' NetBiosName is ' + $netBiosName)
+
+        # Get all schedule tasks that have State of "Ready" and not disabled and "Running"
+        $ScheduledTasks = Get-ScheduledTask | Where-Object { $_.TaskPath -notlike "*\Microsoft\Windows*" -and $_.State -ne "Disabled" -and $_.state -ne "Running" }
```
Comment (jworkmanjc) [inline jumpcloud-ADMU/Powershell/Start-Migration.ps1:1551]: Should we turn this into a function?

`Set-ScheduledTasks -op "Disable" -Tasks $scheduledTasks`
`Set-ScheduledTasks -op "Enable" -Tasks $scheduledTasks`

We have duplicated code in this PR so I'd suggest we move this into a function and determine the operation type with a parameter.

**Review by jworkmanjc — CHANGES_REQUESTED** [review]
I think we should be turning the code in Start-Migration to disable/ enable tasks into a function. We call re-enable tasks twice so it would make sense to just move this to one or two lines rather than the same code twice. 

Question for the team: Do we consider this patch or minor functionality?

File: jumpcloud-ADMU/Powershell/Start-Migration.ps1 (hunk 7 above)
Comment (gweinjc) [reply jumpcloud-ADMU/Powershell/Start-Migration.ps1:1551]: +1 on this

File: jumpcloud-ADMU/Powershell/Start-Migration.ps1 (hunk 6 above)
Comment (kmaranionjc) [reply jumpcloud-ADMU/Powershell/Start-Migration.ps1:1436]: I agree with 2.5, it's new functionality

**Review by gweinjc — APPROVED** [review]

File: ModuleChangelog.md [hunk 8]
```diff
@@ -1,3 +1,13 @@
+## 2.5.0
+
+Release Date: August 28, 2023
+
+#### RELEASE NOTES
+
+```
+* Added disable scheduled task functionality during migration
```
Comment (jworkmanjc) [inline ModuleChangelog.md:8]:
```suggestion
* The ADMU now checks for scheduled tasks before migration and attempts to disable any non-microsoft task. Scheduled tasks which load a user's registry into memory have been reported to have locked a user's registry into memory which will prevent the ADMU from functioning. This release of ADMU will attempt to disable any root level scheduled tasks and will re-enable these tasks after migration or if the ADMU fails to migrate.
```

File: ModuleChangelog.md [hunk 9]
```diff
@@ -1,3 +1,13 @@
...
+  - Disables scheduled task that is on "Ready" state
```
Comment (jworkmanjc) [inline ModuleChangelog.md:9]:
```suggestion
  - Only tasks that are in a "Ready" state will be disabled, currently running tasks are not stopped.
```

**Review by jworkmanjc — APPROVED** [review]
Tested with my windows 11 VM, looks like the expected tasks were disabled, re-enabled after migration. @kmaranionjc we just need to make our changelog a bit more clear before we release but otherwise these changes look good. Thank you for the work here.


Screen Shot 2023-08-29 at 1 37 04 PM
Screen Shot 2023-08-29 at 1 34 57 PM

**Review by gweinjc — APPROVED** [review]
Tested on Win 10, looks like tasks were disabled and then enabled as intended. Nice work!

**Pull request merged**